src/v/ssx/rwlock.h

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include "seastarx.h"
+#include "vlog.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/rwlock.hh>
+#include <seastar/core/timer.hh>
+#include <seastar/util/defer.hh>
+#include <seastar/util/log.hh>
+
+using namespace std::chrono_literals;
+
+namespace ssx {
+
+template<typename F>
+concept ReturnsFutureHolder = requires(F f) {
+    { f() } -> std::same_as<ss::future<ss::rwlock::holder>>;
+};
+
+// Wrapper around seastar::rwlock, meant to be used as a drop-in replacement.
+// Logs an error if the lock has waited a long time to take.
+class logging_rwlock {
+public:
+    using Clock = ss::lowres_clock;
+    using holder_id = int64_t;
+    using resource_str_func = std::function<ss::sstring()>;
+    static constexpr Clock::duration default_log_timeout = 10000ms;
+
+    struct holder {
+    public:
+        holder() = delete;
+        holder(const holder&) = delete;
+
+        holder(logging_rwlock* lock, ss::rwlock::holder&& h, holder_id id)
+          : _lock(lock)
+          , _h(std::move(h))
+          , _id(id) {}
+        holder(holder&& o) noexcept
+          : _lock(std::exchange(o._lock, nullptr))
+          , _h(std::move(o._h))
+          , _id(o._id) {}
+        holder& operator=(holder&& o) noexcept {
+            _lock = std::exchange(o._lock, nullptr);
+            _h = std::move(o._h);
+            _id = o._id;
+            return *this;
+        }
+        ~holder() {
+            if (_lock) {
+                vlog(
+                  _lock->_logger.trace,
+                  "Dropping lock on {} ({})",
+                  _lock->_resource_str_gen(),
+                  _id);
+            }
+        }
+
+    private:
+        logging_rwlock* _lock;
+        ss::rwlock::holder _h;
+        holder_id _id;
+    };
+
+    explicit logging_rwlock(ss::logger& logger)
+      : _logger(logger)
+      , _resource_str_gen([this] { return default_resource_str_gen(); }) {}
+
+    explicit logging_rwlock(ss::logger& logger, resource_str_func fn)
+      : _logger(logger)
+      , _resource_str_gen(std::move(fn)) {}
+
+    ss::sstring default_resource_str_gen() const {
+        return fmt::format("{}", static_cast<const void*>(&_lock));
```

Comment:
i wonder how useful it is to print the addr of the lock vs always requiring the ctor to force the subsystem caller to have the explicit name, so you can actually chase it by grepping the source code. 

very cool idea tho. 

may make sense to have a 2 param ctor as the only ctor, but curious as to why you think the addr of the lock is useful as a default arg.

Reply:
awesome. great utility!